### PR TITLE
feat/channelization: Add channels to instruments

### LINKED ISSUE #478: Decadac reset when restarting the kernel
If you encounter a bug use the following template.
If you have a feature request feel free to freestyle.
### Steps to reproduce
1.  deca.ch0_voltage(1)
2.  restart the kernel
3.  deca.ch0_voltage()

### Expected behaviour
Channel 0 should still be set at 1 V after the kernel has been restarted

### Actual behaviour
Channel 0 is set to 0 V once the kernel has been restarted

### System
**operating system**
Windows 7
**qcodes version**
0.1.2

### PR DESCRIPTION
**Pull request by spauka**
Implements #273

This merge implements channelization in much the way that was discussed in #273, and commits two sample instruments that demonstrate the key features. Briefly, the syntax allows the creation of lists of channels within an instrument and manipulation of those channels. Examples of usage appear at the bottom of the commit.

Changes proposed in this pull request:
- Add InstrumentChannel and ChannelList classes to hold channel classes. Addresses #273.
- Create a driver for the Lakeshore Model336 Temperature controller that uses channels
- Modify the driver for the Harvard DecaDAC to use channels. Fixes #478, #546, #563.

For example:
```pycon
>>> import qcodes as qc
>>> from Harvard.Decadac import Decadac
>>> d = Decadac("DAC", "ASRL4::INSTR")
Connected to Harvard DecaDAC (hw ver: 0, serial: 0) in 0.47s
>>> # Set a single channel
>>> d.channels[0].volt.set(-1)
>>> # Get a single channel
>>> d.channels.volt[0].get()
-1.00006103515625
>>> # Get multiple channels simultaneously
>>> d.channels.volt.get()
(1.00006103515625, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
>>> # Get a subset of channels
>>> d.channels[0:4].volt.get()
(-1.00006103515625, 0.0, 0.0, 0.0)
>>> # Combining different subsets is also possible
>>> fine = d.channels[0:2] + d.channels[4:6]
>>> fine.volt.get()
(-1.00006103515625, 0.0, 0.0, 0.0)
>>> # This also works with measure and loop
>>> qc.Measure(d.channels[0].volt).run()
DataSet:
   location = 'data/2017-04-11/#001_{name}_11-04-36'
   <Type>   | <array_id>           | <array.name> | <array.shape>
   Setpoint | single_set           | single       | (1,)
   Measured | DAC_Slot0_Chan0_volt | volt         | (1,)
acquired at 2017-04-11 11:04:36
```

Things left
 - [x] Channel metadata is not correctly extracted from an instrument snapshot.
 - [x] Test cases need to be written.

@giulioungaretti

### CODE REVIEW

**Comment by giulioungaretti** [comment]
@spauka looking great. Did not check in super detail but ticks all the boxes!

I will tag @MerlinSmiles, since he had strong opinions about this as well!

**Comment by MerlinSmiles** [comment]
@spauka this looks really great! I haven't tested anything on this, but i'm really happy to see this.

What does the 0 in `.get(0)` do here: ?
```
>>> # Get a single channel
>>> d.channels.volt[0].get(0)
```

I have one point that I did not see covered. And that is access to the channels in another notation, (the slicing you show is great, but not the only useful way :)
I'm thinking of the following:
`magnet.x.field.get()`
The dac is a bit special and I see you split it in slots with 4 channels, that would then be:
`d.slot1.ch0.volt.get()`
I see a mismatch with the `dac.channels` you defined in the driver as this list is extended with all the channels. Not sure how to handle this.

One could of course make it on the driver level and define `self.x = self.channels['x']` but if that is possible on one instrument I assume it should consistently be possible on many instruments.

Now im thinking of another syntax also:
````
d.volt.get()
>>>(1,0,0,0...)
```
would also be cool.
But maybe implemented on the driver level as parameters. But again consistency...

@giulioungaretti thanks for tagging, super excited :)

**Comment by spauka** [comment]
@MerlinSmiles For each of your points:

The 0 in `.get(0)` was a typo. Comments above edited.

Regarding different ways of addressing channels, agreed, there should be a way of having "named" channels. I suspect that the solution I am working on to store channel metadata will solve this, new commits coming soon.

However, I think it also makes sense to keep separate the idea of named and numbered channels, i.e. addressing the DAC as dac.chan0 makes less sense than magnet.x.

For the last syntax `d.volt.get()`, sounds good, but it doesn't necessarily make sense to break all channel parameters out in this way. Perhaps we can explicitly define such parameters in the instrument?

Anyway, let me know whether the below commits address your points...

**Comment by MerlinSmiles** [comment]
@spauka totally agree with you!

File: qcodes/instrument/channel.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+        functions (Dict[Function]): All the functions supported by this group of channels
+    """
+
+    def __init__(self, parent, name, chan_type, chan_list=None):
+        super().__init__()
+
+        self._parent = parent
+        self._name = name
+        if not isinstance(chan_type, type) != type or not issubclass(chan_type, InstrumentChannel):
```
Comment (jenshnielsen) [inline qcodes/instrument/channel.py]: As far as I can see the first part of this condition will always be True?

not isinstance(chan_type, type) != type -> not bool != type -> not False.

**Comment by jenshnielsen** [comment]
@spauka What's the status of this. It would be great to have it land soon. It's a much requested feature. I pointed out a small issue above and there is a conflict in the decadac driver now since we merged #552

**Comment by jenshnielsen** [comment]
I'm happy to help writing some of the tests of anything else if you don't have the time

**Comment by spauka** [comment]
@jenshnielsen sorry, got a bit distracted by other projects for the last few weeks but am currently doing work on implementing snapshotting. Any help you can give on tests would be much appreciated.

**Comment by spauka** [comment]
Channel lists should now be snapshotable. This is implemented by creating the concept of a submodule in an instrument, where a submodule is just a class that implements the `Instrument` class.

A side effect of this is that it is now also possible to create sub-instruments, which may be logical groupings of parameters, inside instruments.

To see an example of this, see the `Lakeshore.Model_336` driver, wherein temperature channels may be accessed in two ways:
```py3
therm = Model_336("therm", "...")
# The below two lines access the same parameter
therm.A.temperature()
therm.channels[0].temperature()
# The next statement evaluates to true
therm.A == therm.channels[0]
```

File: qcodes/instrument/channel.py (hunk 1 above)
Comment (spauka) [inline qcodes/instrument/channel.py]: Yep, that's true. Fixed in a future commit :)

**Comment by jenshnielsen** [comment]
@spauka Great I will start working on some tests

File: qcodes/instrument_drivers/Lakeshore/Model_336.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		# or through a channel list.
+		# i.e. Model_336.A.temperature() and Model_336.channels[0].temperature()
+		# refer to the same parameter.
+		channels = ChannelList(self, "TempSensors", SensorChannel, snapshotable=False)
+		for chan_name in ('A', 'B', 'C', 'D'):
+			channel = SensorChannel(self, 'Chan{}'.format(chan_name), chan_name)
+			channels.append(channel)
+			self.add_submodule(chan_name, channel)
+		channels.lock()
+		self.add_submodule("channels", channels)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Lakeshore/Model_336.py]: This fails for me with ```TypeError: Submodules must be subclasses of instrument.``` I guess the check in `add_submodule` should be a bit more permissive?

Comment (spauka) [inline qcodes/instrument_drivers/Lakeshore/Model_336.py]: Yes, you are correct, submodules need to be subclasses of Metadatable, Instrument is too restrictive. This is fixed in 4a3e517

**Comment by jenshnielsen** [comment]
@spauka I have made some changes in my own branch https://github.com/jenshnielsen/qcodes/tree/feat/channelization and I will open a pr against your branch for you to review but in brief I have done the following

* Added some tests for the channels as well as usage within Loops
* Replace a `.` with a `_` in the names generated in line 269. Otherwise datasets cannon be retrieved as member attributes on the returned data from a loop but has to be found via arrays
* overwrite full_name of the MultiParameter, otherwise names have the base instrument applied twice as in `myinstrument_myinstrument_chanA_temperature`
* Add parameters and functions to ChannelList, this was documented but not implemented. I could also be convinced to go the other way and delete this.

**Comment by jenshnielsen** [comment]
PR is here https://github.com/spauka/Qcodes/pull/1

File: qcodes/instrument_drivers/Harvard/Decadac.py [hunk 3]
```diff
@@ -1,266 +1,516 @@
...
+        frac = code/65536.0
+        return (frac * (self.max_val - self.min_val)) + self.min_val
 
+    def _set_slot(self):
+        """
+        Set the active DAC slot
+        """
+        resp = self.ask_raw("B{};".format(self._slot))
+        if int(self._dac_parse(resp)) != self._slot:
+            raise DACException("Unexpected return from DAC when setting slot: {}. DAC slot may not have been set.".format(response))
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: looks like this should be resp

File: qcodes/instrument_drivers/Harvard/Decadac.py [hunk 4]
```diff
@@ -1,266 +1,516 @@
...
 
-    Attributes:
+    def _set_channel(self):
+        """
+        Set the active DAC channel
+        """
+        resp = self.ask_raw("B{};C{};".format(self._slot, self._channel))
+        if resp.strip() != "B{}!C{}!".format(self._slot, self._channel):
+            raise DACException("Unexpected return from DAC when setting channel: {}. DAC channel may not have been set.".format(response))
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: same response to resp

File: qcodes/instrument_drivers/Harvard/Decadac.py [hunk 5]
```diff
@@ -1,266 +1,516 @@
...
+            A dict containing a serial and hardware version
         """
-        switch = {True: 'ON',
-                  False: 'OFF'}
-        mssg = 'Ramp state: ' + switch[self._ramp_state]
-        mssg += '. Ramp time: {:d} ms.'.format(int(self._ramp_time))
-        return mssg
+        self._feature_detect()
+
+        return {serial: self.serial_no, hardware_version: self.version}
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: looks like this is missing some quotes

File: qcodes/instrument_drivers/Harvard/Decadac.py [hunk 6]
```diff
@@ -1,266 +1,516 @@
...
         Args:
-            code (str): The code string from the instrument.
-
-            channel (int): The relevant channel.
+            begin_time (number): time.time() when init started.
+                Default is self._t0, set at start of Instrument.__init__.
         """
+        # start with an empty dict, just in case an instrument doesn't
+        # heed our request to return all 4 fields.
+        t = time.time() - (begin_time or self._t0)
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: This doesn't seem to match the `from time import time` above

File: qcodes/instrument_drivers/Harvard/Decadac.py [hunk 7]
```diff
@@ -1,266 +1,516 @@
...
+        # If we have access to the VERSADAC (slot) EEPROM, we can set the inital
+        # value of the channel.
+        # NOTE: these values will be overwritten by a K3 calibration
+        if self._parent._VERSA_EEPROM_available:
+            _INITIAL_ADDR = [6, 8, 32774, 32776]
+            self.add_parameter("initial_value",
+                get_cmd=partial(self._query_address, _INITIAL_ADDR[self._channel], versa_eeprom=True),
+                get_parser=self._dac_code_to_v,
+                set_cmd=partial(self._write_address, _INITIAL_ADDR[self._channel], versa_eeprom=True),
+                set_parser=set._dac_v_to_code, vals=vals.Numbers(self.min_val, self.max_val))
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: Should probably be self?

File: qcodes/instrument_drivers/Harvard/Decadac.py (hunk 3 above)
Comment (spauka) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: Yep, this is correct. I will run some more tests on the Decadac driver tomorrow and make these changes.

File: qcodes/instrument_drivers/Harvard/Decadac.py (hunk 4 above)
Comment (spauka) [inline qcodes/instrument_drivers/Harvard/Decadac.py]: Agreed again

**Comment by spauka** [comment]
As discussed with @jenshnielsen, I will aim to do some more testing of the Decadac driver in the next day or so, then let's try to port over some more instrument drivers to using channels to ensure that there are no unexpected issues.

- [ ] Additional testing of Decadac driver
- [ ] Port additional drivers to ensure there are no unexpected issues.

**Comment by jenshnielsen** [comment]
In addition to that I would like to add support for ArrayParameters (plus a clean error for MultiParameters) in Channellist slicing

**Comment by jenshnielsen** [comment]
Work in progress support for slicing Array parameters and tests are in https://github.com/jenshnielsen/Qcodes/tree/feat/channelization_with_array

Seems to work good but names are currently incorrect for array and multi parameters saved to dataset via loop. (missing channel prefix in names i.e. this and that instead of ChanA_this and ChanA_that and so on

**Comment by jenshnielsen** [comment]
I pushed some changes to https://github.com/spauka/Qcodes/pull/2 adding support for ArrayParameters

**Comment by jenshnielsen** [comment]
@WilliamHPNielsen is experimenting on porting the QDac driver

**Comment by WilliamHPNielsen** [comment]
I added three commits that:
- Add a channelised QDac driver
- Add a `set` method to MultiChannelInstrumentParameter
- Add a notebook for the channelised QDac

**Comment by WilliamHPNielsen** [comment]
The QDac driver now contains its own subclass of `MultiChannelInstrumentParameter` and a change to `channels.py` that allows for this to happen.

@spauka I think we should merge soon, did you have a chance to look at the Decadac driver?

**Comment by WilliamHPNielsen** [comment]
PEP8 would be nice, but I think we can merge now. @jenshnielsen ?

**Comment by jenshnielsen** [comment]
Yes It would be nice to have https://github.com/spauka/Qcodes/pull/2 merged into this before merging I think

**Comment by spauka** [comment]
I will merge spauka#2 tomorrow (Sydney time) if that's ok with you all, but
I agree with above, let's lock features here and deal with new additions in
future pull requests.

**Comment by jenshnielsen** [comment]
Landed via #640 Thank you very much @spauka

**Pull request closed without merging**